x/wasm/keeper/keeper.go

```diff
@@ -802,22 +758,37 @@ func (k Keeper) setContractInfoExtension(ctx sdk.Context, contractAddr sdk.AccAd
 	return nil
 }
 
-// handleContractResponse processes the contract response
-func (k *Keeper) handleContractResponse(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, res *wasmvmtypes.Response) ([]byte, error) {
-	return k.wasmVMResponseHandler.Handle(ctx, contractAddr, ibcPort, res.Submessages, res.Messages, res.Data)
+// handleContractResponse processes the contract response data by emitting events and sending sub-/messages.
+func (k *Keeper) handleContractResponse(
+	ctx sdk.Context,
+	contractAddr sdk.AccAddress,
+	ibcPort string,
+	subMsg []wasmvmtypes.SubMsg,
+	msgs []wasmvmtypes.CosmosMsg,
+	attrs []wasmvmtypes.EventAttribute,
+	data []byte,
+) ([]byte, error) {
+	attributeGasCost := k.gasRegister.EventCosts(attrs)
+	ctx.GasMeter().ConsumeGas(attributeGasCost, "Custom contract event attributes")
+	// emit all events from this contract itself
+	events := types.ParseEvents(attrs, contractAddr)
+	ctx.EventManager().EmitEvents(events)
+	return k.wasmVMResponseHandler.Handle(ctx, contractAddr, ibcPort, subMsg, msgs, data)
 }
 
-func (k Keeper) gasForContract(ctx sdk.Context) uint64 {
+func (k Keeper) runtimeGasForContract(ctx sdk.Context) uint64 {
 	meter := ctx.GasMeter()
 	if meter.IsOutOfGas() {
 		return 0
 	}
-	remaining := (meter.Limit() - meter.GasConsumedToLimit()) * k.gasMultiplier
-	return remaining
+	if meter.Limit() == 0 { // infinite gas meter with limit=0 and not out of gas
+		return math.MaxUint64
```

Comment:
This was running into an overflow for an infinite gas meter (genesis) before.  Not sure if MaxUint64 makes sense here but it is the highest number available

Reply:
I would ask @webmaster128 what the largest number we can send on the backend.

I am not sure if it is MaxUint64 or MaxInt64 (or something else). But yeah, let's expose what the Rust runtime can handle.

Reply:
`MaxUint64` is good. If it does not work anywhere, it is a bug to be fixed.

I just wonder why we need a special treatment for `0`? Is this behaviour standard in the Cosmos SDK metering?

Reply:
`InfiniteGasMeter` is a bit "special".

You can keep subtracting gas and it never throws and error, just don't query it.

Reply:
https://github.com/cosmos/cosmos-sdk/blob/v0.42.6/store/types/gas.go#L133-L185

Reply:
Alright, that's not my business. Encoding special cases into 0 just always feels like writing C code.

Reply:
Made issue https://github.com/cosmos/cosmos-sdk/issues/9577